services/httpd/handler.go

```diff
@@ -993,8 +1017,437 @@ func fixLiterals(node influxql.Node) {
+		h.httpError(w, fmt.Sprintf("buckets - two retention policies specified: %q differs from %q", rp, brd.Rp), http.StatusBadRequest)
+		return
+	} else if rp == "" {
+		rp = brd.Rp
+
+	}
+	// We only need to validate non-empty
+	// retention policy names
+	if rp != "" && !meta.ValidName(rp) {
+		h.httpError(w, fmt.Sprintf("buckets - retention policy %q: %s", rp, meta.ErrInvalidName.Error()), http.StatusBadRequest)
+		return
+	}
+
+	var dur, sgDur time.Duration
+	if len(brd.RetentionRules) > 0 {
+		dur = time.Second * time.Duration(brd.RetentionRules[0].EverySeconds)
+		sgDur = time.Duration(brd.RetentionRules[0].ShardGroupDurationSeconds) * time.Second
+	} else {
+		dur = meta.DefaultRetentionPolicyDuration
+		// This will get set to default in normalisedShardDuration()
+		// called by CreateRetentionPolicy
+		sgDur = 0
+	}
+	rf := meta.DefaultRetentionPolicyReplicaN
+	spec := meta.RetentionPolicySpec{
+		Name:               rp,
+		Duration:           &dur,
+		ReplicaN:           &rf,
+		ShardGroupDuration: sgDur,
+	}
+	dbi := h.MetaClient.Database(db)
+
+	if h.Config.AuthEnabled {
+		if user == nil {
+			h.httpError(w, fmt.Sprintf("buckets - user is required to to create bucket %q", brd.Name), http.StatusForbidden)
+			return
+		}
+
+		if dbi == nil {
+			if err := h.QueryAuthorizer.AuthorizeCreateDatabase(user); err != nil {
+				h.httpError(w, fmt.Sprintf("buckets - %q is not authorized to create %q: %s", user.ID(), brd.Name, err.Error()), http.StatusForbidden)
+				return
+			}
+		} else {
+			if err := h.QueryAuthorizer.AuthorizeCreateRetentionPolicy(user, db); err != nil {
+				h.httpError(w, fmt.Sprintf("buckets - %q is not authorized to create %q: %s", user.ID(), brd.Name, err.Error()), http.StatusForbidden)
+				return
+			}
+		}
+	}
+
+	if dbi == nil {
+		if _, err = h.MetaClient.CreateDatabaseWithRetentionPolicy(db, &spec); err != nil {
+			h.httpError(w, fmt.Sprintf("buckets - cannot create bucket %q: %s", brd.Name, err.Error()), http.StatusBadRequest)
+			return
+		}
+	} else {
+		if _, err = h.MetaClient.CreateRetentionPolicy(db, &spec, false); err != nil {
+			h.httpError(w, fmt.Sprintf("buckets - cannot create bucket %q: %s", brd.Name, err.Error()), http.StatusBadRequest)
+			return
+		}
+	}
+}
+
+func (h *Handler) serveBucketDeleteV2(w http.ResponseWriter, r *http.Request, user meta.User) {
+	db := r.URL.Query().Get(":db")
+	rp := r.URL.Query().Get(":rp")
+	id := fmt.Sprintf("%s/%s", db, rp)
+
+	if h.Config.AuthEnabled {
+		if user == nil {
+			h.httpError(w, fmt.Sprintf("delete bucket - user is required to delete from database %q", db), http.StatusForbidden)
+			return
+		}
+		if err := h.QueryAuthorizer.AuthorizeCreateRetentionPolicy(user, db); err != nil {
+			h.httpError(w, fmt.Sprintf("buckets - %q is not authorized to modify %q: %s", user.ID(), id, err.Error()), http.StatusForbidden)
+			return
+		}
+	}
+
+	if dbi := h.MetaClient.Database(db); dbi == nil {
```

Comment:
If it is the only retention policy in the DB, should we delete the DB record as well?

Reply:
No, because they may want to create a new retention policy in it.  This API is not a full-service management system for databases and retention policies.